device/src/async_device/mod.rs

```diff
@@ -460,105 +470,136 @@ where
-                Either::Right(_) => Rx1::Timeout(window_duration),
-            }
-        };
-
-        match response {
-            Rx1::Rx(sz) => {
+        let rx_config = self.region.get_rx_config(self.datarate, frame, &Window::_1);
+        #[cfg(feature = "defmt")]
+        trace!("RX1 with config: {}", &rx_config);
+        let expected_payload_air_time_us =
+            rx_config.payload_time_on_air_us(expected_payload_length as u32) + 50_000;
+        #[cfg(feature = "defmt")]
+        trace!("expected payload air time: {}ms", expected_payload_air_time_us / 1_000);
+        match self
+            .rx_with_preamble_deadline(
+                rx_config,
+                rx1_end_delay,
+                expected_payload_air_time_us / 1_000,
+            )
+            .await
+        {
+            Ok((length, _status)) => {
                 self.phy.radio.low_power().await.map_err(Error::Radio)?;
-                return Ok(sz);
+                return Ok(length as usize);
             }
-            Rx1::Timeout(window_duration) => {
-                // If the window duration was the same as the RX2 start delay, we can skip settings the
-                // radio to lower power and arming the the timer
-                if window_duration != rx2_start_delay {
-                    self.phy.radio.low_power().await.map_err(Error::Radio)?;
-                    self.timer.at(rx2_start_delay.into()).await;
-                }
+            Err(err) => {
+                #[cfg(feature = "defmt")]
+                debug!("failed receiving an RX1 response: {}", err);
+                self.phy.radio.low_power().await.map_err(Error::Radio)?;
+                self.timer.at(rx2_start_delay.into()).await; // TODO: what if it's already expired?
             }
         }
 
-        // RX2
-        let response = {
-            // Prepare for RX using correct configuration
-            let rx_config = self.region.get_rx_config(self.datarate, frame, &Window::_2);
-            let window_duration = self.phy.radio.get_rx_window_duration_ms();
-
-            // Pass the full radio buffer slice to RX
-            self.phy.radio.setup_rx(rx_config).await.map_err(Error::Radio)?;
-            let rx_fut = self.phy.radio.rx(self.radio_buffer.as_raw_slice());
-            let timeout_fut = self.timer.delay_ms(window_duration.into());
-
-            pin_mut!(rx_fut);
-            pin_mut!(timeout_fut);
-            // Wait until either RX is complete or if we've reached window close
-            match select(rx_fut, timeout_fut).await {
-                // RX is complete!
-                Either::Left((Ok((sz, _q)), _)) => Ok(sz),
-                // Timeout or other RX error.
-                _ => Err(Error::RxTimeout),
-            }
-        };
+        let rx_config = self.region.get_rx_config(self.datarate, frame, &Window::_2);
+        #[cfg(feature = "defmt")]
+        trace!("RX2 with config: {}", &rx_config);
```

Comment:
maybe put these traces behind an additional feature-gate? seems excessive

Reply:
if DEFMT_LOG is lower than "trace" at compile-time, the whole macro evaluates to nothing